# Ordenador con el que se ha hecho el experimento

Características:
    
   * Intel® Core™ i9-12900H CPU @ 2.40GHz × 4 
   * 32 GiB RAM DDR4
   * 1T SSD m.2
   * Sistema Operativo: Ubuntu
   * Empleamos el archivo mencionado en el manual. (DATOS DEL 2018)
   
Ya podemos intuir que no siendo SSD no se obtengan buenos tiempos.

NOTA IMPORTANTE:

   * AMD (up to 2.4 GHz)
   * 16 GB DDR3
   * 1000 GB HDD con 600 GB libres
   * Sistema Operativo: Windows

# Librerías Vaex y Dask documentación

* Vaex

https://vaex.io/
    
https://vaex.io/docs/index.html

* Dask

https://www.dask.org/
    
https://tutorial.dask.org/

* Dataset taxis de Nueva York:

https://www1.nyc.gov/site/tlc/about/tlc-trip-record-data.page

# Pandas

In [3]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [23]:
%%time
import pandas as pd

CPU times: user 4 µs, sys: 1e+03 ns, total: 5 µs
Wall time: 9.06 µs


Tiempos: 

   * Con 1TB HDD sobre Windows: 1.95 segundos (alguna vez hasta 12 segundos) 
   * Con 1TB SSD sobre Linux: 571 milisegundos
   * Con 1TB SSD m.2 Linux: 519 ms

In [17]:
#%%time
#df_pandas = pd.read_csv("yellow_tripdata.csv")
#df_pandas.head()

****

In [ ]:
# read_parquet AÑADIDO PARA LA NUEVA VERSIÓN:

In [24]:
%%time
df_pandas = pd.read_parquet("yellow_tripdata_2018-01.parquet")
df_pandas.head()

CPU times: user 3.79 s, sys: 2.94 s, total: 6.73 s
Wall time: 2.03 s


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2018-01-01 00:21:05   2018-01-01 00:24:23                1   
1         1  2018-01-01 00:44:55   2018-01-01 01:03:05                1   
2         1  2018-01-01 00:08:26   2018-01-01 00:14:21                2   
3         1  2018-01-01 00:20:22   2018-01-01 00:52:51                1   
4         1  2018-01-01 00:09:18   2018-01-01 00:27:06                2   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            0.5           1                  N            41            24   
1            2.7           1                  N           239           140   
2            0.8           1                  N           262           141   
3           10.2           1                  N           140           257   
4            2.5           1                  N           246           239   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          4.5    0.5      0.5        0.00           0.0   
1             2         14.0    0.5      0.5        0.00           0.0   
2             1          6.0    0.5      0.5        1.00           0.0   
3             2         33.5    0.5      0.5        0.00           0.0   
4             1         12.5    0.5      0.5        2.75           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    0.3          5.80                   NaN          NaN  
1                    0.3         15.30                   NaN          NaN  
2                    0.3          8.30                   NaN          NaN  
3                    0.3         34.80                   NaN          NaN  
4                    0.3         16.55                   NaN          NaN

In [9]:
#Transformar el dataframe para poder trabajar con él.
#df_pandas.to_csv("yellow_tripdata_2.csv")

****

Tiempos:

   * Con 1TB HDD sobre Windows: 1 minuto 6 segundos (a veces algo más)
   * Con 1TB SSD sobre Linux:   56.6 segundos
   * Con 250 SSD sobre linux:   21.6 segundos
   * Con 1TB SSD m.2 Linux: 2.64 s

In [25]:
print('Number of Rows: ',len(df_pandas.index))
print("Number of Columns: " + str(len(df_pandas.axes[1]))) 

Number of Rows:  8760687
Number of Columns: 19


In [26]:
%%time
df_pandas.describe()

CPU times: user 3.56 s, sys: 613 ms, total: 4.17 s
Wall time: 4.19 s


VendorID  passenger_count  trip_distance    RatecodeID  \
count  8.760687e+06     8.760687e+06   8.760687e+06  8.760687e+06   
mean   1.560978e+00     1.606807e+00   2.804022e+00  1.039545e+00   
std    4.962678e-01     1.258420e+00   6.412050e+01  4.450619e-01   
min    1.000000e+00     0.000000e+00   0.000000e+00  1.000000e+00   
25%    1.000000e+00     1.000000e+00   9.100000e-01  1.000000e+00   
50%    2.000000e+00     1.000000e+00   1.550000e+00  1.000000e+00   
75%    2.000000e+00     2.000000e+00   2.840000e+00  1.000000e+00   
max    2.000000e+00     9.000000e+00   1.894838e+05  9.900000e+01   

       PULocationID  DOLocationID  payment_type   fare_amount         extra  \
count  8.760687e+06  8.760687e+06  8.760687e+06  8.760687e+06  8.760687e+06   
mean   1.644579e+02  1.627270e+02  1.310613e+00  1.224443e+01  3.246882e-01   
std    6.635990e+01  7.031145e+01  4.817808e-01  1.168321e+01  4.502555e-01   
min    1.000000e+00  1.000000e+00  1.000000e+00 -4.500000e+02 -4.469000e+01   
25%    1.160000e+02  1.130000e+02  1.000000e+00  6.000000e+00  0.000000e+00   
50%    1.620000e+02  1.620000e+02  1.000000e+00  9.000000e+00  0.000000e+00   
75%    2.340000e+02  2.340000e+02  2.000000e+00  1.350000e+01  5.000000e-01   
max    2.650000e+02  2.650000e+02  4.000000e+00  8.016000e+03  6.000000e+01   

            mta_tax    tip_amount  tolls_amount  improvement_surcharge  \
count  8.760687e+06  8.760687e+06  8.760687e+06           8.760687e+06   
mean   4.975066e-01  1.818759e+00  3.026157e-01           2.996307e-01   
std    4.333281e-02  2.486375e+00  1.738184e+00           1.442748e-02   
min   -5.000000e-01 -8.880000e+01 -1.500000e+01          -3.000000e-01   
25%    5.000000e-01  0.000000e+00  0.000000e+00           3.000000e-01   
50%    5.000000e-01  1.360000e+00  0.000000e+00           3.000000e-01   
75%    5.000000e-01  2.350000e+00  0.000000e+00           3.000000e-01   
max    4.549000e+01  4.417100e+02  9.507000e+02           1.000000e+00   

       total_amount  congestion_surcharge  airport_fee  
count  8.760687e+06                  12.0         12.0  
mean   1.549109e+01                   2.5          0.0  
std    1.419546e+01                   0.0          0.0  
min   -4.503000e+02                   2.5          0.0  
25%    8.300000e+00                   2.5          0.0  
50%    1.130000e+01                   2.5          0.0  
75%    1.662000e+01                   2.5          0.0  
max    8.016800e+03                   2.5          0.0

Tiempos:

   * Con 1TB HDD sobre Windows: 11.7 segundos
   * Con 1TB SSD sobre Linux:   7.76 segundos
   * Con 250 SSD sobre linux:   4.81 segundos
   * Con 1TB SSD m.2 Linux: 2.3 s

In [27]:
%%time
df_pandas.fare_amount.value_counts()

CPU times: user 79.1 ms, sys: 10.6 ms, total: 89.7 ms
Wall time: 88.1 ms


6.00       473270
5.50       465207
6.50       461959
7.00       446414
5.00       433292
            ...  
30.60           1
2409.00         1
168.88          1
201.50          1
33.96           1
Name: fare_amount, Length: 1714, dtype: int64

Tiempos:

   * Con 1TB HDD sobre Windows: 555 milisegundos
   * Con 1TB SSD sobre Linux:   313 milisegundos
   * Con 250 SSD sobre linux:   213 milisegundos
   * Con 1TB SSD m.2 Linux: 44 ms

In [28]:
%%time
# casi 9 millones de filas en este caso
len(df_pandas), df_pandas.shape

CPU times: user 20 µs, sys: 1e+03 ns, total: 21 µs
Wall time: 26.2 µs


(8760687, (8760687, 19))

Tiempos:

   * Con 1TB HDD sobre Windows: 0 nanosegundos
   * Con 1TB SSD sobre Linux:   46.7 microsegundos
   * Con 250 SSD sobre linux:   21 microsegundos
   * Con 1TB SSD m.2 Linux: 18 microsegundos

In [29]:
%%time
df_pandas.tail()

CPU times: user 171 µs, sys: 33 µs, total: 204 µs
Wall time: 188 µs


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
8760682         1  2018-01-31 23:21:35   2018-01-31 23:34:20                2   
8760683         1  2018-01-31 23:35:51   2018-01-31 23:38:57                1   
8760684         2  2018-01-31 23:28:00   2018-01-31 23:37:09                1   
8760685         2  2018-01-31 23:24:40   2018-01-31 23:25:28                1   
8760686         2  2018-01-31 23:28:16   2018-01-31 23:28:38                1   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
8760682           2.80           1                  N           158   
8760683           0.60           1                  N           163   
8760684           2.95           1                  N            74   
8760685           0.00           1                  N             7   
8760686           0.00           1                  N             7   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
8760682           163             1         12.0    0.5      0.5        2.65   
8760683           162             1          4.5    0.5      0.5        1.15   
8760684            69             2         10.5    0.5      0.5        0.00   
8760685           193             2          0.0    0.0      0.0        0.00   
8760686           193             2          0.0    0.0      0.0        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
8760682           0.0                    0.3         15.95   
8760683           0.0                    0.3          6.95   
8760684           0.0                    0.3         11.80   
8760685           0.0                    0.0          0.00   
8760686           0.0                    0.0          0.00   

         congestion_surcharge  airport_fee  
8760682                   NaN          NaN  
8760683                   NaN          NaN  
8760684                   NaN          NaN  
8760685                   NaN          NaN  
8760686                   NaN          NaN

Tiempos:

   * Con 1TB HDD sobre Windows: 0 nanosegundos
   * Con 1TB SSD sobre Linux:   215 microsegundos
   * Con 250 SSD sobre linux:   327 microsegundos
   * Con 1TB SSD m.2 Linux: 753 microsegundos

# DASK

In [11]:
#pip install dask

Note: you may need to restart the kernel to use updated packages.


https://docs.dask.org/en/stable/install.html

In [47]:
%%time
import dask.dataframe as dd

CPU times: user 8 µs, sys: 1e+03 ns, total: 9 µs
Wall time: 13.8 µs


Tiempos:

   * Con 1TB HDD sobre Windows: 4.13 segundos (promedio 6 segundos, a veces 8 segundos)
   * Con 1TB SSD sobre Linux:   846 milisegundos
   * Con 250 SSD sobre linux:   792 milisegundos
   * Con 1TB SSD m.2 Linux: 134 milisegundos

In [48]:
%%time
df_dask = dd.read_csv("yellow_tripdata_2.csv",
                      assume_missing=True)
df_dask.head()

CPU times: user 1.44 s, sys: 306 ms, total: 1.74 s
Wall time: 1.76 s


Unnamed: 0  VendorID tpep_pickup_datetime tpep_dropoff_datetime  \
0         0.0       1.0  2018-01-01 00:21:05   2018-01-01 00:24:23   
1         1.0       1.0  2018-01-01 00:44:55   2018-01-01 01:03:05   
2         2.0       1.0  2018-01-01 00:08:26   2018-01-01 00:14:21   
3         3.0       1.0  2018-01-01 00:20:22   2018-01-01 00:52:51   
4         4.0       1.0  2018-01-01 00:09:18   2018-01-01 00:27:06   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0              1.0            0.5         1.0                  N   
1              1.0            2.7         1.0                  N   
2              2.0            0.8         1.0                  N   
3              1.0           10.2         1.0                  N   
4              2.0            2.5         1.0                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0          41.0          24.0           2.0          4.5    0.5      0.5   
1         239.0         140.0           2.0         14.0    0.5      0.5   
2         262.0         141.0           1.0          6.0    0.5      0.5   
3         140.0         257.0           2.0         33.5    0.5      0.5   
4         246.0         239.0           1.0         12.5    0.5      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        0.00           0.0                    0.3          5.80   
1        0.00           0.0                    0.3         15.30   
2        1.00           0.0                    0.3          8.30   
3        0.00           0.0                    0.3         34.80   
4        2.75           0.0                    0.3         16.55   

   congestion_surcharge  airport_fee  
0                   NaN          NaN  
1                   NaN          NaN  
2                   NaN          NaN  
3                   NaN          NaN  
4                   NaN          NaN

Tiempos:

   * Con 1TB HDD sobre Windows: 5.9 segundos
   * Con 1TB SSD sobre Linux:   1.94 segundos
   * Con 250 SSD sobre linux:   1.88 segundos
   * Con 1TB SSD m.2 Linux: 550 milisegundos

****

In [ ]:
# Leer el archivo parquet:

In [49]:
%%time
df_dask = dd.read_parquet("yellow_tripdata_2018-01.parquet",
                      assume_missing=True)
df_dask.head()

CPU times: user 3.33 s, sys: 1.98 s, total: 5.31 s
Wall time: 5.65 s


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2018-01-01 00:21:05   2018-01-01 00:24:23                1   
1         1  2018-01-01 00:44:55   2018-01-01 01:03:05                1   
2         1  2018-01-01 00:08:26   2018-01-01 00:14:21                2   
3         1  2018-01-01 00:20:22   2018-01-01 00:52:51                1   
4         1  2018-01-01 00:09:18   2018-01-01 00:27:06                2   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            0.5           1                  N            41            24   
1            2.7           1                  N           239           140   
2            0.8           1                  N           262           141   
3           10.2           1                  N           140           257   
4            2.5           1                  N           246           239   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          4.5    0.5      0.5        0.00           0.0   
1             2         14.0    0.5      0.5        0.00           0.0   
2             1          6.0    0.5      0.5        1.00           0.0   
3             2         33.5    0.5      0.5        0.00           0.0   
4             1         12.5    0.5      0.5        2.75           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    0.3          5.80                   NaN          NaN  
1                    0.3         15.30                   NaN          NaN  
2                    0.3          8.30                   NaN          NaN  
3                    0.3         34.80                   NaN          NaN  
4                    0.3         16.55                   NaN          NaN

****

In [50]:
%%time
df_dask.describe()

CPU times: user 115 ms, sys: 4.55 ms, total: 119 ms
Wall time: 116 ms


Dask DataFrame Structure:
              VendorID passenger_count trip_distance RatecodeID PULocationID DOLocationID payment_type fare_amount    extra  mta_tax tip_amount tolls_amount improvement_surcharge total_amount congestion_surcharge airport_fee
npartitions=1                                                                                                                                                                                                                   
               float64         float64       float64    float64      float64      float64      float64     float64  float64  float64    float64      float64               float64      float64              float64     float64
                   ...             ...           ...        ...          ...          ...          ...         ...      ...      ...        ...          ...                   ...          ...                  ...         ...
Dask Name: describe-numeric, 68 graph layers

Tiempos:

   * Con 1TB HDD sobre Windows: 465 milisegundos
   * Con 1TB SSD sobre Linux:   298 milisegundos
   * Con 250 SSD sobre linux:   220 milisegundos
   * Con 1TB SSD m.2 Linux: 56.7 milisegundos

In [51]:
%%time
df_dask.describe().compute()

CPU times: user 14 s, sys: 8.89 s, total: 22.9 s
Wall time: 14.9 s


VendorID  passenger_count  trip_distance    RatecodeID  \
count  8.760687e+06     8.760687e+06   8.760687e+06  8.760687e+06   
mean   1.560978e+00     1.606807e+00   2.804022e+00  1.039545e+00   
std    4.962678e-01     1.258420e+00   6.412050e+01  4.450619e-01   
min    1.000000e+00     0.000000e+00   0.000000e+00  1.000000e+00   
25%    1.000000e+00     1.000000e+00   9.100000e-01  1.000000e+00   
50%    2.000000e+00     1.000000e+00   1.550000e+00  1.000000e+00   
75%    2.000000e+00     2.000000e+00   2.840000e+00  1.000000e+00   
max    2.000000e+00     9.000000e+00   1.894838e+05  9.900000e+01   

       PULocationID  DOLocationID  payment_type   fare_amount         extra  \
count  8.760687e+06  8.760687e+06  8.760687e+06  8.760687e+06  8.760687e+06   
mean   1.644579e+02  1.627270e+02  1.310613e+00  1.224443e+01  3.246882e-01   
std    6.635990e+01  7.031145e+01  4.817808e-01  1.168321e+01  4.502555e-01   
min    1.000000e+00  1.000000e+00  1.000000e+00 -4.500000e+02 -4.469000e+01   
25%    1.160000e+02  1.130000e+02  1.000000e+00  6.000000e+00  0.000000e+00   
50%    1.620000e+02  1.620000e+02  1.000000e+00  9.000000e+00  0.000000e+00   
75%    2.340000e+02  2.340000e+02  2.000000e+00  1.350000e+01  5.000000e-01   
max    2.650000e+02  2.650000e+02  4.000000e+00  8.016000e+03  6.000000e+01   

            mta_tax    tip_amount  tolls_amount  improvement_surcharge  \
count  8.760687e+06  8.760687e+06  8.760687e+06           8.760687e+06   
mean   4.975066e-01  1.818759e+00  3.026157e-01           2.996307e-01   
std    4.333281e-02  2.486375e+00  1.738184e+00           1.442748e-02   
min   -5.000000e-01 -8.880000e+01 -1.500000e+01          -3.000000e-01   
25%    5.000000e-01  0.000000e+00  0.000000e+00           3.000000e-01   
50%    5.000000e-01  1.360000e+00  0.000000e+00           3.000000e-01   
75%    5.000000e-01  2.350000e+00  0.000000e+00           3.000000e-01   
max    4.549000e+01  4.417100e+02  9.507000e+02           1.000000e+00   

       total_amount  congestion_surcharge  airport_fee  
count  8.760687e+06                  12.0         12.0  
mean   1.549109e+01                   2.5          0.0  
std    1.419546e+01                   0.0          0.0  
min   -4.503000e+02                   2.5          0.0  
25%    8.300000e+00                   2.5          0.0  
50%    1.130000e+01                   2.5          0.0  
75%    1.662000e+01                   2.5          0.0  
max    8.016800e+03                   2.5          0.0

Tiempos:

   * Con 1TB HDD sobre Windows: 50 segundos (57.5 segundos otra de las veces)
   * Con 1TB SSD sobre Linux:   36 segundos
   * Con 250 SSD sobre linux:   40.6 segundos
   * Con 1TB SSD m.2 Linux: 9.43 segundos

In [52]:
%%time
df_dask.fare_amount.value_counts()

CPU times: user 1.57 ms, sys: 74 µs, total: 1.64 ms
Wall time: 1.6 ms


Dask Series Structure:
npartitions=1
    int64
      ...
Name: fare_amount, dtype: int64
Dask Name: value-counts-agg, 4 graph layers

Tiempos:

   * Con 1TB HDD sobre Windows: 5 milisegundos
   * Con 1TB SSD sobre Linux:   3.25 milisegundos
   * Con 250 SSD sobre linux:   1.65 milisegundos
   * Con 1TB SSD m.2 Linux: 2.08 milisegundos

In [53]:
%%time
df_dask.fare_amount.value_counts().compute()

CPU times: user 267 ms, sys: 92.9 ms, total: 360 ms
Wall time: 380 ms


6.00       473270
5.50       465207
6.50       461959
7.00       446414
5.00       433292
            ...  
60.06           1
60.30           1
60.53           1
60.55           1
8016.00         1
Name: fare_amount, Length: 1714, dtype: int64

Tiempos:

   * Con 1TB HDD sobre Windows: 13.1 segundos
   * Con 1TB SSD sobre Linux:   5.45 segundos
   * Con 250 SSD sobre linux:   12.5 segundos
   * Con 1TB SSD m.2 Linux: 149 milisegundos

In [19]:
pip install vaex

  Using cached vaex-4.14.0-py3-none-any.whl (4.7 kB)
  Using cached vaex_server-0.8.1-py3-none-any.whl (23 kB)
  Using cached vaex_jupyter-0.8.0-py3-none-any.whl (43 kB)
  Using cached vaex_hdf5-0.13.0-py3-none-any.whl (16 kB)
  Using cached vaex_ml-0.18.0-py3-none-any.whl (58 kB)
  Using cached vaex_astro-0.9.2-py3-none-any.whl (20 kB)
  Using cached vaex_viz-0.5.4-py3-none-any.whl (19 kB)
  Using cached astropy-5.1-cp38-cp38-macosx_10_9_x86_64.whl (6.9 MB)
  Using cached bqplot-0.12.36-py2.py3-none-any.whl (1.2 MB)
  Using cached ipympl-0.9.2-py2.py3-none-any.whl (510 kB)
  Using cached ipyvuetify-1.8.4-py2.py3-none-any.whl (11.7 MB)
  Using cached ipyvolume-0.5.2-py2.py3-none-any.whl (2.9 MB)
  Using cached ipyleaflet-0.17.2-py3-none-any.whl (3.7 MB)
  Using cached xarray-2022.10.0-py3-none-any.whl (947 kB)
  Using cached numba-0.56.3.tar.gz (2.4 MB)
  Preparing metadata (setup.py) ... done
  Using cached pythreejs-2.4.1-py3-none-any.whl (3.3 MB)
  Using cached ipyvue-1.8.0-py2.py3-

  Using cached ipydatawidgets-4.3.2-py2.py3-none-any.whl (271 kB)


  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [875 lines of output]
      /private/var/folders/jw/lg6ll5ln5sx8l3dbpj7412480000gn/T/pip-install-pvl6hi_s/numba_24bce256d2e74b119e1e0ca287365317/setup.py:144: DeprecationWarning:
      
        `numpy.distutils` is deprecated since NumPy 1.23.0, as a result
        of the deprecation of `distutils` itself. It will be removed for
        Python >= 3.12. For older Python versions it will remain present.
        It is recommended to use `setuptools < 60.0` for those Python versions.
        For more details, see:
          https://numpy.org/devdocs/reference/distutils_status_migration.html
      
      
        import numpy.distutils.misc_util as np_misc
      TBB not found
      OpenMP not found
      running bdist_wheel
      running build
      got version from file /private/var/folders/jw/lg6ll5ln5sx8l3dbpj7412480000gn/T/pip-install-pvl6hi_s/numba_24bce256d2e74b119

Failed to build numba
  Running setup.py install for numba ... error
  error: subprocess-exited-with-error
  
  × Running setup.py install for numba did not run successfully.
  │ exit code: 1
  ╰─> [53 lines of output]
      /private/var/folders/jw/lg6ll5ln5sx8l3dbpj7412480000gn/T/pip-install-pvl6hi_s/numba_24bce256d2e74b119e1e0ca287365317/setup.py:144: DeprecationWarning:
      
        `numpy.distutils` is deprecated since NumPy 1.23.0, as a result
        of the deprecation of `distutils` itself. It will be removed for
        Python >= 3.12. For older Python versions it will remain present.
        It is recommended to use `setuptools < 60.0` for those Python versions.
        For more details, see:
          https://numpy.org/devdocs/reference/distutils_status_migration.html
      
      
        import numpy.distutils.misc_util as np_misc
      TBB not found
      OpenMP not found
      running install
      running build
      got version from file /private/var/folders/jw/lg6ll5

# VAEX desde un CSV (1ª forma)

In [37]:
%%time
# https://pypi.org/project/vaex/
#pip install vaex
import vaex

CPU times: user 8 µs, sys: 17 µs, total: 25 µs
Wall time: 27.9 µs


Tiempos:

   * Con 1TB HDD sobre Windows: 7.64 segundos (en otro intento fueron 14.7 segundos)
   * Con 1TB SSD sobre Linux:   5.14 segundos
   * Con 250 SSD sobre linux:   1.85 segundos
   * Con 1TB SSD m.2 Linux: 365 milisegundos

In [38]:
%%time
# Necesito añadir convert=True para que me convierta .csv en .HDF5
# Default chunk_size for converting is 5 million rows, 
# which corresponds to around 1Gb memory on an example of NYC Taxi dataset.
df_vaex = vaex.from_csv("yellow_tripdata_2.csv", 
                        convert=True, chunk_size = 5_000_000)

df_vaex

ModuleNotFoundError: No module named 'vaex.hdf5'

Tiempos:

   * Con 1TB HDD sobre Windows: 3 minutos 24 segundos (alguna vez mucho menos)
   * Con 1TB SSD sobre Linux:   1 minuto 10 segundos
   * Con 250 SSD sobre linux:   1 minuto 8 segundos

# VAEX desde HDF5

In [18]:
%%time
# https://pypi.org/project/vaex/
#pip install vaex
import vaex

CPU times: user 5 µs, sys: 1e+03 ns, total: 6 µs
Wall time: 8.82 µs


In [20]:
%%time
vaex_hdf5 = vaex.open("yellow_tripdata_2.csv.hdf5")
vaex_hdf5

[10/21/22 12:22:17] ERROR    error opening 'yellow_tripdata_2.csv.hdf5'                             ]8;id=562965;file:///Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vaex/__init__.py\__init__.py]8;;\:]8;id=104118;file:///Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vaex/__init__.py#262\262]8;;\
                             Traceback (most recent call last):                                                    
                               File                                                                                
                             "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-                
                             packages/vaex/__init__.py", line 235, in open                                         
                                 ds = vaex.dataset.open(path, fs_options=fs_options, fs=fs,                        
                             **kwargs)                                                                             
                               File                                                                                
                             "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-                
                             packages/vaex/dataset.py", line 88, in open                                           
                                 raise IOError(f'Cannot open {path} nobody knows how to read it.')                 
                             OSError: Cannot open yellow_tripdata_2.csv.hdf5 nobody knows how to                   
                             read it.                                                                              

OSError: Cannot open yellow_tripdata_2.csv.hdf5 nobody knows how to read it.

Tiempos:

   * Con 1TB HDD sobre Windows: 5.11 segundos (alguna ocasión fueron 57 milisegundos)
   * Con 1TB SSD sobre Linux:   40.1 milisegundos
   * Con 250 SSD sobre linux:   26.5 milisegundos
   * Con 1TB SSD m.2 Linux: 9.53 milisegundos

In [21]:
%%time
vaex_hdf5.describe()

NameError: name 'vaex_hdf5' is not defined

Tiempos:

   * Con 1TB HDD sobre Windows: 9.56 segundos (a veces bastante más)
   * Con 1TB SSD sobre Linux:   4.56 segundos
   * Con 250 SSD sobre linux:   4.29 segundos
   * Con 1TB SSD m.2 Linux: 8.1 segundos

In [10]:
%%time
vaex_hdf5.fare_amount.value_counts()

NameError: name 'vaex_hdf5' is not defined

Tiempos:

   * Con 1TB HDD sobre Windows: 1.62 segundos (No ejecuta correctamente alguna vez)
   * Con 1TB SSD sobre Linux:   114 milisegundos
   * Con 250 SSD sobre linux:   336 milisegundos
   * Con 1TB SSD m.2 Linux: 373 milisegundos

# VAEX (2ª forma)

In [ ]:
%%time
# https://pypi.org/project/vaex/
# pip install vaex
import vaex

Tiempos:

   * Con 1TB HDD sobre Windows: 7.28 segundos
   * Con 1TB SSD sobre Linux:   19.6 microsegundos
   * Con 250 SSD sobre linux:   13.1 microsegundos

In [43]:
%%time
df_vaex_3 = vaex.from_csv("yellow_tripdata_2.csv")
df_vaex_3

CPU times: user 20.5 s, sys: 7.84 s, total: 28.4 s
Wall time: 26.2 s


#          Unnamed: 0    VendorID    tpep_pickup_datetime    tpep_dropoff_datetime    passenger_count    trip_distance    RatecodeID    store_and_fwd_flag    PULocationID    DOLocationID    payment_type    fare_amount    extra    mta_tax    tip_amount    tolls_amount    improvement_surcharge    total_amount    congestion_surcharge    airport_fee
0          0             1           2018-01-01 00:21:05     2018-01-01 00:24:23      1                  0.5              1             N                     41              24              2               4.5            0.5      0.5        0.0           0.0             0.3                      5.8             nan                     nan
1          1             1           2018-01-01 00:44:55     2018-01-01 01:03:05      1                  2.7              1             N                     239             140             2               14.0           0.5      0.5        0.0           0.0             0.3                      15.3            nan                     nan
2          2             1           2018-01-01 00:08:26     2018-01-01 00:14:21      2                  0.8              1             N                     262             141             1               6.0            0.5      0.5        1.0           0.0             0.3                      8.3             nan                     nan
3          3             1           2018-01-01 00:20:22     2018-01-01 00:52:51      1                  10.2             1             N                     140             257             2               33.5           0.5      0.5        0.0           0.0             0.3                      34.8            nan                     nan
4          4             1           2018-01-01 00:09:18     2018-01-01 00:27:06      2                  2.5              1             N                     246             239             1               12.5           0.5      0.5        2.75          0.0             0.3                      16.55           nan                     nan
...        ...           ...         ...                     ...                      ...                ...              ...           ...                   ...             ...             ...             ...            ...      ...        ...           ...             ...                      ...             ...                     ...
8,760,682  8760682       1           2018-01-31 23:21:35     2018-01-31 23:34:20      2                  2.8              1             N                     158             163             1               12.0           0.5      0.5        2.65          0.0             0.3                      15.95           nan                     nan
8,760,683  8760683       1           2018-01-31 23:35:51     2018-01-31 23:38:57      1                  0.6              1             N                     163             162             1               4.5            0.5      0.5        1.15          0.0             0.3                      6.95            nan                     nan
8,760,684  8760684       2           2018-01-31 23:28:00     2018-01-31 23:37:09      1                  2.95             1             N                     74              69              2               10.5           0.5      0.5        0.0           0.0             0.3                      11.8            nan                     nan
8,760,685  8760685       2           2018-01-31 23:24:40     2018-01-31 23:25:28      1                  0.0              1             N                     7               193             2               0.0            0.0      0.0        0.0           0.0             0.0                      0.0             nan                     nan
8,760,686  8760686       2           2018-01-31 23:28:16     2018-01-31 23:28:38      1                  0.0              1             N                     7               193             2               0.0            0.0      0.0        0.0        

Tiempos:

   * Con 1TB HDD sobre Windows: 1 minuto 23 segundos
   * Con 1TB SSD sobre Linux:   54.6 segundos
   * Con 250 SSD sobre linux:   47.9 segundos
   * Con 1TB SSD m.2 Linux: 7.92 segundos

In [44]:
%%time

df_vaex_3.describe()

CPU times: user 7.96 s, sys: 579 ms, total: 8.54 s
Wall time: 2.67 s


Unnamed: 0           VendorID tpep_pickup_datetime  \
data_type           int64              int64               string   
count             8760687            8760687              8760687   
NA                      0                  0                    0   
mean            4380343.0  1.560978037452999                   --   
std        2528992.498871           0.496268                   --   
min                     0                  1                   --   
max               8760686                  2                   --   

          tpep_dropoff_datetime     passenger_count       trip_distance  \
data_type                string               int64             float64   
count                   8760687             8760687             8760687   
NA                            0                   0                   0   
mean                         --  1.6068074341658365  2.8040221012347226   
std                          --             1.25842           64.120494   
min                          --                   0                 0.0   
max                          --                   9           189483.84   

                   RatecodeID store_and_fwd_flag        PULocationID  \
data_type               int64             string               int64   
count                 8760687            8760687             8760687   
NA                          0                  0                   0   
mean       1.0395453004998352                 --  164.45790906580729   
std                  0.445062                 --           66.359897   
min                         1                 --                   1   
max                        99                 --                 265   

                 DOLocationID        payment_type         fare_amount  \
data_type               int64               int64             float64   
count                 8760687             8760687             8760687   
NA                          0                   0                   0   
mean       162.72696718876043  1.3106125124662027  12.244425999924466   
std                 70.311441            0.481781           11.683206   
min                         1                   1              -450.0   
max                       265                   4              8016.0   

                         extra              mta_tax          tip_amount  \
data_type              float64              float64             float64   
count                  8760687              8760687             8760687   
NA                           0                    0                   0   
mean       0.32468824648112593  0.49750661106828725  1.8187593050640622   
std                   0.450255             0.043333            2.486375   
min                     -44.69                 -0.5               -88.8   
max                       60.0                45.49              441.71   

                  tolls_amount improvement_surcharge        total_amount  \
data_type              float64               float64             float64   
count                  8760687               8760687             8760687   
NA                           0                     0                   0   
mean       0.30261571837938245    0.2996306796425908  15.491088333739462   
std                   1.738184              0.014427           14.195456   
min                      -15.0                  -0.3              -450.3   
max                      950.7                   1.0              8016.8   

          congestion_surcharge airport_fee  
data_type              float64     float64  
count                       12          12  
NA                     8760675     8760675  
mean                       2.5         0.0  
std                        0.0         0.0  
min                        2.5         0.0  
max                        2.5         0.0

In [45]:
%%time
df_vaex_3.export('yellow_tripdata_2018-01.csv.hdf5')

ModuleNotFoundError: No module named 'vaex.hdf5'

Tiempos:

   * Con 1TB HDD sobre Windows: 1 minuto 19 segundos (53.8 segundos en otro intento)
   * Con 1TB SSD sobre Linux:   9.73 segundos
   * Con 250 SSD sobre linux:   7.13 segundos
   * Con 1TB SSD m.2 Linux: 1.36 segundos

In [46]:
%%time
df_vaex_4 = vaex.open('yellow_tripdata_2018-01.csv.hdf5')
df_vaex_4

[10/21/22 12:32:28] ERROR    error opening 'yellow_tripdata_2018-01.csv.hdf5'                       ]8;id=830925;file:///Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vaex/__init__.py\__init__.py]8;;\:]8;id=773624;file:///Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vaex/__init__.py#262\262]8;;\
                             Traceback (most recent call last):                                                    
                               File                                                                                
                             "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-                
                             packages/vaex/__init__.py", line 235, in open                                         
                                 ds = vaex.dataset.open(path, fs_options=fs_options, fs=fs,                        
                             **kwargs)                                                                             
                               File                                                                                
                             "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-                
                             packages/vaex/dataset.py", line 88, in open                                           
                                 raise IOError(f'Cannot open {path} nobody knows how to read it.')                 
                             OSError: Cannot open yellow_tripdata_2018-01.csv.hdf5 nobody knows how                
                             to read it.                                                                           

OSError: Cannot open yellow_tripdata_2018-01.csv.hdf5 nobody knows how to read it.

Tiempos:

   * Con 1TB HDD sobre Windows: 9.28 segundos (5.83 segundos en otro intento, 84 milisegundos en otro)
   * Con 1TB SSD sobre Linux:   61.1 milisegundos
   * Con 250 SSD sobre linux:   142 milisegundos
   * Con 1TB SSD m.2 Linux: 6.31 milisegundos

 # CONCLUSIÓN

Cuando se genera con Vaex el hdf5 lo ejecuta mas rápido que sin él, sino tarda más que el resto.